db/versions/0009-migration.sql

```diff
@@ -0,0 +1,29 @@
+begin
+  -- lock this table before reading from it, to prevent loss of concurrent
+  -- updates when the table is dropped.  Note that this may lead to concurrent
+  -- updates failing; the important thing is that they not succeed without
+  -- taking effect.  Failed updates will be retried.
+  lock table cache_purges_entities;
+
+  create table cache_purges
+  as
+    select
+      (value ->> 'provisionerId')::text as provisioner_id,
+      (value ->> 'workerType')::text as worker_type,
+      (value ->> 'cacheName')::text as cache_name,
+      (value ->> 'before')::timestamptz as before,
+      (value ->> 'expires')::timestamptz as expires,
+      etag
+    from cache_purges_entities;
+  alter table cache_purges add primary key (provisioner_id, worker_type, cache_name);
+  alter table cache_purges
+    alter column provisioner_id set not null,
+    alter column worker_type set not null,
+    alter column cache_name set not null,
+    alter column before set not null,
+    alter column expires set not null;
```

Comment:
so if we do have null data in there now, this will fail and then we can roll-back from there? I ask because it seems like we're seeing some schema validation errors on output these days which hints at maybe we have missing data somewhere. I'm not worried and 100% confident that this will be fine, just want to understand the process we would go through.

Reply:
> so if we do have null data in there now, this will fail and then we can roll-back from there? 

Yes. It will fail without consequences other than the time spent migrating.

> I ask because it seems like we're seeing some schema validation errors on output these days which hints at maybe we have missing data somewhere

Interesting! Is this for a known or unknown subset of tables?

Reply:
https://sentry.prod.mozaws.net/operations/taskcluster-firefox-ci/issues/7969678/?query=is:unresolved%20validation in particular is what I was thinking of. no real ideas if it is caused by db issues or something else.

I think we shouldn't worry about it too much especially since this will fail gracefully.